1110-i386-txt-Add-Intel-TXT-core-implementation.patch

```diff
@@ -0,0 +1,1038 @@
++                   grub_uint32_t mem_type)
++{
++  grub_uint64_t mtrr_def_type;
++  grub_uint64_t mtrr_cap;
++  union mtrr_physbase_t mtrr_physbase;
++  union mtrr_physmask_t mtrr_physmask;
++  grub_uint32_t vcnt, pages_in_range;
++  unsigned long ndx, base_v;
++  int i = 0, j, num_pages, mtrr_s;
++
++  /* Disable all fixed MTRRs, set default type to UC */
++  mtrr_def_type = grub_rdmsr (GRUB_MSR_X86_MTRR_DEF_TYPE);
++  mtrr_def_type &= ~(GRUB_MSR_X86_MTRR_ENABLE_FIXED | GRUB_MSR_X86_DEF_TYPE_MASK);
++  mtrr_def_type |= GRUB_MTRR_MEMORY_TYPE_UC;
++  grub_wrmsr (GRUB_MSR_X86_MTRR_DEF_TYPE, mtrr_def_type);
++
++  /* Initially disable all variable MTRRs (we'll enable the ones we use) */
++  mtrr_cap = grub_rdmsr (GRUB_MSR_X86_MTRRCAP);
++  vcnt = (mtrr_cap & GRUB_MSR_X86_VCNT_MASK);
++
++  for ( ndx = 0; ndx < vcnt; ndx++ )
++    {
++      mtrr_physmask.raw = grub_rdmsr (GRUB_MSR_X86_MTRR_PHYSMASK0 + ndx*2);
++      mtrr_physmask.v = 0;
++      grub_wrmsr (GRUB_MSR_X86_MTRR_PHYSMASK0 + ndx*2, mtrr_physmask.raw);
++    }
++
++  /* Map all AC module pages as mem_type */
++  num_pages = GRUB_PAGE_UP(size) >> GRUB_PAGE_SHIFT;
++
++  grub_dprintf ("slaunch", "setting MTRRs for acmod: base=%p, size=%x, num_pages=%d\n",
++           base, size, num_pages);
++
++  /* Each VAR MTRR base must be a multiple if that MTRR's Size */
++  base_v = (unsigned long)base;
++  /* MTRR size in pages */
++  mtrr_s = 1;
++
++  while ( (base_v & 0x01) == 0 )
++    {
++      i++;
++      base_v = base_v >> 1;
++    }
++
++  for (j = i - 12; j > 0; j--)
++     mtrr_s = mtrr_s*2; /* mtrr_s = mtrr_s << 1 */
++
++  grub_dprintf ("slaunch", "The maximum allowed MTRR range size=%d Pages \n", mtrr_s);
++
++  ndx = 0;
++
++  while ( num_pages >= mtrr_s )
++    {
++      mtrr_physbase.raw = grub_rdmsr (GRUB_MSR_X86_MTRR_PHYSBASE0 + ndx*2);
++      mtrr_physbase.base = ((unsigned long)base >> GRUB_PAGE_SHIFT) &
++	                     SINIT_MTRR_MASK;
++      mtrr_physbase.type = mem_type;
++      grub_wrmsr (GRUB_MSR_X86_MTRR_PHYSBASE0 + ndx*2, mtrr_physbase.raw);
++
++      mtrr_physmask.raw = grub_rdmsr (GRUB_MSR_X86_MTRR_PHYSMASK0 + ndx*2);
++      mtrr_physmask.mask = ~(mtrr_s - 1) & SINIT_MTRR_MASK;
++      mtrr_physmask.v = 1;
++      grub_wrmsr (GRUB_MSR_X86_MTRR_PHYSMASK0 + ndx*2, mtrr_physmask.raw);
++
++      base += (mtrr_s * GRUB_PAGE_SIZE);
++      num_pages -= mtrr_s;
++      ndx++;
++      if ( ndx == vcnt )
++          return grub_error (GRUB_ERR_BAD_DEVICE,
++			     N_("exceeded number of var MTRRs when mapping range"));
++    }
++
++  ndx = 0;
++
++  while ( num_pages > 0 )
++    {
++      /* Set the base of the current MTRR */
++      mtrr_physbase.raw = grub_rdmsr (GRUB_MSR_X86_MTRR_PHYSBASE0 + ndx*2);
++      mtrr_physbase.base = ((unsigned long)base >> GRUB_PAGE_SHIFT) &
++                            SINIT_MTRR_MASK;
++      mtrr_physbase.type = mem_type;
++      grub_wrmsr (GRUB_MSR_X86_MTRR_PHYSBASE0 + ndx*2, mtrr_physbase.raw);
```

Comment:
This should be a utility function.  The repetition makes the math very difficult to follow.